Выполнил практическое задание спринта №15

MaximusDenisov-YP opened the pull request.

MaximusDenisov-YP committed:
Начал выполнение практического задания спринта №15

- Добавил соответствующие аннотации сущностям приложения.
- Добавил использование JPA в приложение, вынес хранение в PostgreSQL.

MaximusDenisov-YP committed:
Описал шаблон управления Booking

- Реализовал функционал User и Item.
- Добавил новое бизнес исключение NotAvailableException.java, соответственно изменив GlobalExceptionHandler.java
- Добавлены мапперы к каждой сущности.
- Удалены имплементации репозиториев для хранения данных приложения в оперативной памяти.

MaximusDenisov-YP committed:
Описал шаблон управления Booking

- Реализовал функционал User и Item.
- Добавил новое бизнес исключение NotAvailableException.java, соответственно изменив GlobalExceptionHandler.java
- Добавлены мапперы к каждой сущности.
- Удалены имплементации репозиториев для хранения данных приложения в оперативной памяти.

MaximusDenisov-YP committed:
Merge remote-tracking branch 'origin/add-bookings' into add-bookings

# Conflicts:
#	src/main/java/ru/practicum/shareit/booking/dto/BookingDto.java
#	src/main/java/ru/practicum/shareit/booking/entity/Booking.java
#	src/main/java/ru/practicum/shareit/booking/mapper/BookingMapper.java
#	src/main/java/ru/practicum/shareit/booking/repository/BookingRepository.java
#	src/main/java/ru/practicum/shareit/booking/service/BookingService.java
#	src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java

MaximusDenisov-YP committed:
Выполнил практическое задание спринта №15

MaximusDenisov-YP committed:
Исправил замечания по CodeStyle.

MaximusDenisov-YP committed:
Исправил замечания по Check and Build application

avfyodorov commented on src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -15,18 +17,30 @@ public class BookingController {
+
+    @GetMapping("/{bookingId}")
+    public BookingDto getBookingById(
+            @Valid @Positive @PathVariable Long bookingId,
```
`            @Valid @Positive @PathVariable Long bookingId,`

`@Valid` на примитивных типах не работает,

.-----------------------------
Валидация данных в Spring Boot
https://struchkov.dev/blog/ru/spring-boot-validation/
.---------------------------

Проверка данных — Java & Spring Validation

https://habr.com/ru/post/424819/

avfyodorov replied:
`          @Positive`

`           ........ @Positive @PathVariable Long bookingId,`

Обратите внимание, что необходимо добавить `@Validated` в контроллер на уровне класса, чтобы проверять параметры метода. В этом случае аннотация `@Validated` устанавливается на уровне класса, даже если она присутствует на методах.

avfyodorov commented on src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -15,18 +17,30 @@ public class BookingController {
-        return bookingService.createBooking(bookingDto);
+    public BookingDto createNewBooking(
+            @Valid @RequestBody BookingCreateDto bookingDto,
+            @Valid @Positive @RequestHeader("X-Sharer-User-Id") Long userId) {
```
Здесь и дальше аналогичные комментарии по поводу аннотаций для проверки данных.

avfyodorov commented on src/main/java/ru/practicum/shareit/booking/mapper/BookingMapper.java:
```diff
@@ -0,0 +1,45 @@
+import java.time.format.DateTimeFormatter;
+import java.util.function.Function;
+
+public class BookingMapper implements Function<Booking, BookingDto> {
```
Принято максимально избавляться от шаблонного кода, чтобы не перегружать и так большие и объемные проекты. Для маппинга сущностей часто используется библиотека Mapstruct, которая позволяет автоматически настраивать маппинг из одной сущности в другую с помощью описания интерфейсов и аннотаций.
Вот здесь есть отличный гайд для старта работы с библиотекой.
https://web.archive.org/web/20240619130052/https://rukovodstvo.net/posts/id_799/

//  уже нет. только из архива  https://rukovodstvo.net/posts/id_799/

Что такое Mapstruct и как правильно настроить его для модульного тестирования в SpringBoot приложениях
https://javarush.com/groups/posts/3698-chto-takoe-mapstruct-i-kak-praviljhno-nastroitjh-ego-dlja-moduljhnogo-testirovanija-v-springboo

При наличии свободного времени можно поизучать и попробовать прикрутить в проект
Необязательно, просто для информации . 
.------------

Выглядит, конечно... не очень. 🙂 Будет время, всё-таки посмотрите в сторону MapStruct
Необязательно, на Ваше усмотрение.

MaximusDenisov-YP replied:
@avfyodorov , большое спасибо за совет, обычно я всегда доделываю исправления до конца по вашим замечаниям, но в этот раз попытался внедрить этот подход, идею понял, сильно сократил кол-во кода, но это повлекло слишком много борьбы с зависимостями, а у меня горят учебные сроки.. 
Я взял себе на заметку номер 0, изучу и в последствии применю и буду применять именно MapStruct.

avfyodorov commented on src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -35,6 +37,13 @@ public ItemDto createNewItem(@Valid @RequestBody ItemDto itemDto, @RequestHeader
     }
 
+    @PostMapping("/{itemId}/comment")
+    public CommentDto createNewComment(@Valid @RequestBody CommentDto commentDto,
```
Как правильно проставить аннотации. 🙂
Если есть ограничительные аннотации на примитивные входные параметры, например - `Positive `, то над классом нужно ставить аннотацию `Validated `

@Validated              <-- !!!!!!!!!!!!!!!!
public class FilmController {
....................
public Film get(@PathVariable @Positive <----!!!!!!!!!!!!!!!!   Long id,..

Если же нужно проверить входной дто объект (NewCommentDto ..), то здесь пригодится аннотация Valid - 

public Film create(@RequestBody @Valid  <--- !!!!!!!!!!!!! Film film) {
                                               ..................

avfyodorov reviewed (commented):
Добрый день, Максим! 

Хорошая работа, всего несколько небольших уточнений.

MaximusDenisov-YP committed:
Исправления по замечаниям от ревьюера.

- Исправил некорректные проверки параметров на вход, добавил аннотации включающие остальные проверки.
- Попытка добавить использование @Mapper от MapStruct
- Rollback MapStruct, так как пришлось очень многие вещи изменять, взял на вооружение, в последствии применю.

avfyodorov reviewed (commented):
Добрый день, Максим!

Замечаний нет.
Работа принята.
